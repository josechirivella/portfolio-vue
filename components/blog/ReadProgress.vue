<template>
  <div class="read-progress-container">
    <span
      :class="{ 'with-shadow': 'shadow' }"
      :style="{
        width: readProgress + '%',
        color: color,
        height: height,
        opacity: opacity,
        backgroundColor: color,
      }"
      class="read-progress-bar"
    />
  </div>
</template>

<script lang="ts" setup>
defineProps({
  color: {
    type: String,
    default: 'var(--p-primary-color)',
  },
  height: {
    type: String,
    default: '4px',
  },
  opacity: {
    type: [String, Number],
    default: 1,
  },
  shadow: {
    type: Boolean,
    default: true,
  },
});
const readProgress = ref(0);

onMounted(() => {
  window.addEventListener('scroll', updateReadProgress);
});

onBeforeUnmount(() => {
  window.removeEventListener('scroll', updateReadProgress);
});

function updateReadProgress(): void {
  readProgress.value = currentScrollPosition(window.scrollY);
}

function currentScrollPosition(position: number): number {
  return (
    (position /
      (document.body.clientHeight - document.documentElement.clientHeight)) *
    100
  );
}
</script>

<style lang="scss" scoped>
.read-progress-container {
  width: 100%;
  background-color: transparent;
  position: fixed;
  display: block;
  z-index: 9999;
  top: 0;
  left: 0;

  .read-progress-bar {
    display: block;
    width: 0;
    height: inherit;

    &.with-shadow {
      box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.3);
    }
  }
}
</style>
