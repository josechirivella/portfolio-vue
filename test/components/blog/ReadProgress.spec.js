import { beforeAll, describe, expect, test, vi } from 'vitest';
import { shallowMount } from '@vue/test-utils';
import ReadProgress from '~/app/components/blog/ReadProgress.vue';

// Mock window methods and properties
Object.defineProperty(window, 'addEventListener', { value: vi.fn() });
Object.defineProperty(window, 'removeEventListener', { value: vi.fn() });

// Mock document properties for calculating scroll progress
Object.defineProperty(document.body, 'clientHeight', {
  value: 1000,
  configurable: true,
});
Object.defineProperty(document.documentElement, 'clientHeight', {
  value: 500,
  configurable: true,
});

describe('ReadProgress', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = shallowMount(ReadProgress, {});
  });

  test('should render component', () => {
    expect(wrapper.vm).toBeTruthy();
    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  describe('Props', () => {
    test('should have default props', () => {
      expect(wrapper.props('color')).toBe('var(--p-primary-color)');
      expect(wrapper.props('height')).toBe('4px');
      expect(wrapper.props('opacity')).toBe(1);
      expect(wrapper.props('shadow')).toBe(true);
    });

    test('should apply custom props', async () => {
      await wrapper.setProps({
        color: 'red',
        height: '8px',
        opacity: 0.5,
        shadow: false,
      });

      expect(wrapper.props('color')).toBe('red');
      expect(wrapper.props('height')).toBe('8px');
      expect(wrapper.props('opacity')).toBe(0.5);
      expect(wrapper.props('shadow')).toBe(false);
    });
  });

  describe('Read progress calculation', () => {
    test('currentScrollPosition should calculate correct percentage', () => {
      // Mock document dimensions for test
      // When scrollY is 0, progress should be 0%
      expect(wrapper.vm.currentScrollPosition(0)).toBe(0);

      // When scrollY is halfway, progress should be 50%
      expect(wrapper.vm.currentScrollPosition(250)).toBe(50);

      // When scrollY is at max, progress should be 100%
      expect(wrapper.vm.currentScrollPosition(500)).toBe(100);
    });

    test('updateReadProgress should update readProgress value', async () => {
      // Initial value
      expect(wrapper.vm.readProgress).toBe(0);

      // Mock scrollY by spying on the property
      const scrollYSpy = vi
        .spyOn(window, 'scrollY', 'get')
        .mockReturnValue(250);
      await wrapper.vm.updateReadProgress();

      // Check that readProgress was updated
      expect(wrapper.vm.readProgress).toBe(50);

      scrollYSpy.mockRestore();
    });
  });

  describe('Lifecycle hooks', () => {
    test('should add event listener on mount', () => {
      expect(window.addEventListener).toHaveBeenCalledWith(
        'scroll',
        wrapper.vm.updateReadProgress,
      );
    });

    test('should remove event listener before unmount', () => {
      wrapper.unmount();
      expect(window.removeEventListener).toHaveBeenCalledWith(
        'scroll',
        wrapper.vm.updateReadProgress,
      );
    });
  });
});
