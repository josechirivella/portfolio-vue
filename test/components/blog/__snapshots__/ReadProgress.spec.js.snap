// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReadProgress > should render component 1`] = `
<div
  class="read-progress-container"
  data-v-38d9cd48=""
>
  <span
    class="with-shadow read-progress-bar"
    data-v-38d9cd48=""
    style="width: 0%; color: var(--p-primary-color); height: 4px; opacity: 1; background-color: var(--p-primary-color);"
  />
</div>
`;
